### ISSUE TITLE
Checkpoint Aggregation Pattern - Snakemake running one sample at a time

### ISSUE THREAD

**Issue by elsherbini** [snakemake/snakemake#16]
I often make rules that might produce an empty output file. I then have some aggregate step that wants to take all non-empty samples and do something to them.

It appears the current checkpoint functionality doesn't support this fully - snakemake runs one sample at a time and then reevaluates the DAG each time rather than running all samples simultaneously through the checkpoint rule.

Here is a minimal reproducible example that shows the behavior:

`‌``import os
SAMPLES = ["a","b","c","d"]

rule first:
    input: "aggregate_output"

checkpoint before_aggregate:
    output: "{sample}.txt"
    run:
        with open(output[0], "w") as f:
            if wildcards.sample=="a":
                pass
            else:
                f.write("something\n")

def get_aggregate_inputs(wildcards):
    return ["{}.txt".format(s) s in SAMPLES if os.path.getsize(checkpoints.before_aggregate.get(sample=s).output[0])>0]

rule aggregate:
    input:
        get_aggregate_inputs
    output: touch("aggregate_output")
```
It appears checkpoints.before_aggregate.get(sample=s) raises the error for the first sample it tries, so it goes ahead and runs the necessary steps for that one sample. It then tries again and it raises the error for the second sample, etc.

The behavior I’d want with a checkpoint like this is:

- the first time the input function runs, figure out all the checkpoint outputs it depends on and run those.

 - then rerun the input function after all checkpoints have been completed.

**Comment by elsherbini** [comment]
This is achievable by making a dummy rule that takes all the outputs from `maybe_empty`, writes out a list of non empty files, and then using that as input to aggregate. But it's pretty clunky. Is there a smarter way for snakemake to handle this internally to reduce this type of boilerplate?
```
import os
SAMPLES = ["a","b","c","d"]

rule target:
    input: "output/aggregate_output"

rule maybe_empty:
    output: "output/{sample}.txt"
    run:
        with open(output[0], "w") as f:
            if wildcards.sample=="a":
                pass
            else:
                f.write("something\n")

checkpoint write_out_non_empty:
    input: expand("output/{sample}.txt", sample=SAMPLES)
    output: "output/non_empty_files"
    run:
        with open(output[0], "w") as f:
            f.write("\n".join([in_file for in_file in input if os.path.getsize(in_file)>0]))

def get_aggregate_input(wildcards):
    with open(checkpoints.write_out_non_empty.get(**wildcards).output[0], "r") as f:
      non_empty_files = [l.strip() for l in f.read_lines]
    return non_empty_files

rule aggregate:
    input:
        get_aggregate_input
    output: "output/aggregate_output"
    run:
        with open(output[0], "w") as f:
            f.write("\n".join(input))
```

**Comment by troycomi** [comment]
I think checkpoints are working properly here and *really* you don't want to change the execution of the DAG based on input, you want to change what you consider as "input" while running all your rules.  Instead of checkpoints, use all your files as input and then a params function to select only inputs which are non-empty:

```
import os
SAMPLES = ["a","b","c","d"]

rule target:
    input: "output/aggregate_output"

rule maybe_empty:
    output: "output/{sample}.txt"
    run:
        with open(output[0], "w") as f:
            if wildcards.sample=="a":
                pass
            else:
                f.write("something\n")

def get_nonempty_files(wildcards, input):
    return [i for i in input if os.path.getsize(i)>0]

rule aggregate:
    input:
        expand("output/{sample}.txt", sample=SAMPLES)
    output: "output/aggregate_output"
    params: nonempty=get_nonempty_files
    run:
        with open(output[0], "w") as f:
            f.write("\n".join(params.nonempty))
```

The nice thing about the function is you can reuse it throughout your workflow as it is rule-agnostic.  You should also handle cases where you have no non-empty files but I assume you have the covered anyways.  Hope that's smart enough! :smile:

**Comment by elsherbini** [comment]
I like this params trick, I think that could be useful. I think in my case I do want to change the DAG though.

For instance, I want to assemble some genomes, but sometimes there aren't enough reads in a given sample after filtering. Then I want to do some quality checks on the completed genomes, and pull out a specific genes and make an alignments of those genes.

![empty_files_snakemake](https://user-images.githubusercontent.com/2695357/68690067-668fcc80-053f-11ea-8b73-3758d5e6dc1f.png)

sample d has no reads after the step 1 - i'd like to not run any of the subsequent steps for that sample.

sample a doesn't have a hit to one of the genes in step 4. I'd like the alignment for gene 1 to use only sample b and c, and the alignment for gene 2 to use a, b, and c.

Each step will raise an error if I try to run it with empty files.

I could write a wrapper around each step so it can take empty files as input and write an empty file as output, and then wait for the aggregation step (alignment in this case) to use the params trick you have above. But that's a lot of if statements, and I'd rather not submit jobs to the queue that are just going to touch a file.

**Comment by troycomi** [comment]
Great graphic and clear explanation.

This is a different question from the above.  Let me think about it a bit more and see if something clicks.

Just to confirm, step 6 is an aggregation of step 5 into two parts, one with 2 inputs the other with 3?

**Comment by elsherbini** [comment]
> This is a different question from the above. Let me think about it a bit more and see if something clicks.

I agree, the initial example was too pared down to show the whole picture.

> Just to confirm, step 6 is an aggregation of step 5 into two parts, one with 2 inputs the other with 3?

Step 6 might be a rule like the following (with the issue being that sometimes samples need to be abandoned as described above):

```
rule step_6:
  input: expand("step_5_out_{sample}_{{id}}.txt", sample=["a","b","c","d"])
  output: "step_6_out_{id}.txt"
```

**Comment by troycomi** [comment]
I eventually ended up trying the checkpoint write_non_empty but had trouble getting my checkpoint changes to propagate through all the rules.  If anyone else wants to play with the example, here's an **incorrect** workflow with most of the logic for the above dag:
```
samples = 'a b c d'.split()
parts = 'a b'.split()

rule all:
    input: 'output/agg/a', 'output/agg/b'

def get_agg_input(wildcards):
    return expand(f'output/5/{{sample}}_5_{wildcards.part}',
                  sample=samples)

rule step6:
    input: get_agg_input
    output: 'output/agg/{part}'
    shell:
        '''
        cat {input} > {output}
        echo in rule 6 >> {output}
        '''
    
rule step5:
    input: 'output/4/{sample}_4_{part}'
    output: 'output/5/{sample}_5_{part}'
    shell:
        '''
        cat {input} > {output}
        echo in rule 5 {wildcards.sample} >> {output}
        '''

rule step4:
    input: 'output/3/{sample}_3'
    output: expand('output/4/{{sample}}_4_{part}', part=('a', 'b'))
    shell:
        # fail on sample a, part a for this step
        '''
        if [[ {wildcards.sample} == "a" ]]; then
            touch {output[0]}
        else
            cat {input} > {output[0]}
            echo in rule 4 {wildcards.sample} >> {output[0]}
        fi
        cat {input} > {output[1]}
        echo in rule 4 {wildcards.sample} >> {output[1]}
        '''

rule step3:
    input: 'output/2/{sample}_2'
    output: 'output/3/{sample}_3'
    shell:
        '''
        cat {input} > {output}
        echo in rule 3 {wildcards.sample} >> {output}
        '''

rule step2:
    input: 'output/1/{sample}_1'
    output: 'output/2/{sample}_2'
    shell:
        # fail on sample d for this step
        '''
        if [[ {wildcards.sample} == "d" ]]; then
            touch {output}
        else
            cat {input} > {output}
            echo in rule 2 {wildcards.sample} >> {output}
        fi
        '''

rule step1:
    output: 'output/1/{sample}_1'
    shell:
        'echo starting rule 1 {wildcards.sample} > {output}'
```
With expected outputs of:
```
==> agg/a <==
starting rule 1 b
in rule 2 b
in rule 3 b
in rule 4 b
in rule 5 b
starting rule 1 c
in rule 2 c
in rule 3 c
in rule 4 c
in rule 5 c
in rule 6

==> agg/b <==
starting rule 1 a
in rule 2 a
in rule 3 a
in rule 4 a
in rule 5 a
starting rule 1 b
in rule 2 b
in rule 3 b
in rule 4 b
in rule 5 b
starting rule 1 c
in rule 2 c
in rule 3 c
in rule 4 c
in rule 5 c
in rule 6
```

In terms of the blocking checkpoint issue in the first post, have you tried using multiprocessing for the list comprehension?  I have a feeling it would be a disaster but if you are out of ideas it would be worth spending a little time trying.  Alternatively, you could break up the workflow with subworkflows which would produce the non-empty file lists as an intermediate output.  Your main snakefile would take that output resume processing from there.  I get the feeling subworkflows will be more robust than checkpoints.

Sorry I couldn't be of more help!

**Comment by elsherbini** [comment]
Thanks a lot for your time and effort. @troycomi 

I think you are right, subworkflows might be a more elegant solution to this than checkpoints. I think it'll still require "dummy rules" that write out which samples have worked and which have been abandoned, but I think that would be useful to have anyway.

**Comment by cademirch** [comment]
Snakemake has evolved substantially since this was opened. If you believe this issue still applies to a recent version of Snakemake (9.x+), please reopen with updated details. Thanks!
